### PR TITLE
[context] Broaden the host type of context controllers to be compatible with Lit 2

### PR DESCRIPTION
**Pull request by justinfagnani**
This adds `lit@^2.7.5` and `@lit/reactive-element@^1.6.2` as dependencies of context, and builds the tests with those packages mapped in in a separate tsconfig.

That highlighted the type error in the controllers, which I fixed by changing the host type to `ReactiveControllerHost & HTMLElement`.

### CODE REVIEW

**Comment by justinfagnani** [comment]
So the type errors the first commit creates look like this:

<details>
  <summary>Click to see the build output</summary>

```ts
> @lit/context@1.0.0 build:ts:lit2-tests
> wireit

src/lib/decorators/consume.ts:57:29 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(host: ReactiveElement, options: Options<{ __context__: ValueType; }>): ContextConsumer<{ __context__: ValueType; }, ReactiveElement>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.
  Overload 2 of 2, '(host: ReactiveElement, context: { __context__: unknown; }, callback?: ((value: unknown, dispose?: (() => void) | undefined) => void) | undefined, subscribe?: boolean | undefined): ContextConsumer<...>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.

57         new ContextConsumer(this, {
                               ~~~~


src/lib/decorators/consume.ts:70:31 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(host: ReactiveElement, options: Options<{ __context__: ValueType; }>): ContextConsumer<{ __context__: ValueType; }, ReactiveElement>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.
  Overload 2 of 2, '(host: ReactiveElement, context: { __context__: unknown; }, callback?: ((value: unknown, dispose?: (() => void) | undefined) => void) | undefined, subscribe?: boolean | undefined): ContextConsumer<...>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.

70           new ContextConsumer(element, {
                                 ~~~~~~~


src/lib/decorators/provide.ts:62:53 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(host: ReactiveElement, options: Options<{ __context__: ValueType; }>): ContextProvider<{ __context__: ValueType; }>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.
      Types of property 'renderRoot' are incompatible.
        Type 'DocumentFragment | HTMLElement' is not assignable to type 'ShadowRoot | HTMLElement'.
          Type 'DocumentFragment' is not assignable to type 'ShadowRoot | HTMLElement'.
            Type 'DocumentFragment' is missing the following properties from type 'ShadowRoot': delegatesFocus, host, mode, onslotchange, and 11 more.
  Overload 2 of 2, '(host: ReactiveElement, context: { __context__: unknown; }, initialValue?: unknown): ContextProvider<{ __context__: unknown; }>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.

62         controllerMap.set(this, new ContextProvider(this, {context}));
                                                       ~~~~


src/lib/decorators/provide.ts:81:58 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(host: ReactiveElement, options: Options<{ __context__: ValueType; }>): ContextProvider<{ __context__: ValueType; }>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.
  Overload 2 of 2, '(host: ReactiveElement, context: { __context__: unknown; }, initialValue?: unknown): ContextProvider<{ __context__: unknown; }>', gave the following error.
    Argument of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/packages/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement' is not assignable to parameter of type 'import("/Users/justinfagnani/Projects/Lit/lit-monorepo/node_modules/lit2/node_modules/@lit/reactive-element/development/reactive-element", { assert: { "resolution-mode": "import" } }).ReactiveElement'.

81           controllerMap.set(element, new ContextProvider(element, {context}));
                                                            ~~~~~~~



Found 4 errors in 2 files.

Errors  Files
     2  src/lib/decorators/consume.ts:57
     2  src/lib/decorators/provide.ts:62
```

</details>

Notice that the errors are in the _definitions_ of the decorators, not at the use site of the decorators. This is a bit unexpected to me, since the decorators accept a `ReactiveElement` interface that should have incompatible renderRoot declarations between Lit 2 and Lit 3.

**Comment by justinfagnani** [comment]
With the help of @rictic yesterday we figure out the the error I hit on lit.dev only occurs when you have an installation like:

```
lit.dev@0.0.0 /Users/justinfagnani/Projects/Lit/2.0/lit.dev
├─┬ lit-dev-content@0.0.0 -> ./packages/lit-dev-content
│ ├─┬ @lit/context@1.0.0
│ │ └── @lit/reactive-element@1.6.3
│ └─┬ lit@3.0.0
│   ├── @lit/reactive-element@2.0.0
│   └─┬ lit-element@4.0.0
│     └── @lit/reactive-element@2.0.0 deduped
```

So the use site (top-level) is using `@lit/reactive-element@2.0.0` and context is using `@lit/reactive-element@1.6.3`. In this case `renderRoot` including DocumentFragment (in 2.0) is not assignable to `renderRoot` including ShadowRoot (in 1.6).

If you swap the versions so that context is using 2.0 and the use site is using 1.6, then it type-checks fine because ShadowRoot _is_ assignable to DocumentFragment.

Testing context with a different version for the test code is relatively easy - it's just a new tsconfig for tests - testing with a different version for context itself is hard because you need to get the build of the tests to resolve `@lit/reactive-element` in _context_ to a different location, and I didn't figure out how to do that yet without a real node_modules install. Any remapping of paths in the tsconfigs for the tests applies to the tests, and we need a remapping that applies only to the library files, not tests.

I don't know how worth it it is to include a test like this. This is probably a corner case on installation, and there's an available if sometimes cumbersome fix of deduping npm installations. We could apply the fix and just not include these tests, and repeat if we get reports of similar errors externally.

File: packages/context/lit2/lit2-scope/package.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "@lit-internal/context-lit2-scope",
```
Comment (augustjk) [inline packages/context/lit2/lit2-scope/package.json:2]: maybe this also needs `"type": "module"`. build seems to complain that files here are commonjs?

**Comment by augustjk** [comment]
> I don't know how worth it it is to include a test like this. This is probably a corner case on installation, and there's an available if sometimes cumbersome fix of deduping npm installations. We could apply the fix and just not include these tests, and repeat if we get reports of similar errors externally.

I agree with this sentiment.

File: packages/context/src/lib/controllers/context-provider.ts [hunk 2]
```diff
@@ -39,6 +39,17 @@ export interface Options<C extends Context<unknown, unknown>> {
...
+/**
+ * Generates a public interface type that removes private and protected fields.
+ * This allows accepting otherwise compatible versions of the type (e.g. from
+ * multiple copies of the same package in `node_modules`).
+ */
+type Interface<T> = {
+  [K in keyof T]: T[K];
+};
+
+type ReactiveElementHost = Interface<Omit<ReactiveElement, 'renderRoot'>>;
```
Comment (augustjk) [inline packages/context/src/lib/controllers/context-provider.ts:51]: Any reason why this can't also just be `ReactiveContorllerHost & HTMLElement` like for the `ContextConsumer`? Or maybe the other way around to use this for `ContextConsumer`?

Also wdyt about upstreaming this type into `reactive-element` so other decorators can use it too?

Comment (justinfagnani) [reply packages/context/src/lib/controllers/context-provider.ts]: `ReactiveContorllerHost & HTMLElement` works

**Review by augustjk — APPROVED** [review]

File: .changeset/two-spiders-walk.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit/context': patch
+---
+
+Broaden the host type of context controllers to be compatible with Lit 2
```
Comment (augustjk) [inline .changeset/two-spiders-walk.md:5]:
```suggestion
Broaden the host type of context controllers and decorators to be compatible with Lit 2
```
